lib/ecto/type.ex

```diff
@@ -644,45 +644,48 @@ defmodule Ecto.Type do
       :error
 
   """
-  @spec cast(t, term) :: {:ok, term} | {:error, keyword()} | :error
-  def cast({:parameterized, type, params}, value), do: type.cast(value, params)
-  def cast({:in, _type}, nil), do: :error
-  def cast(_type, nil), do: {:ok, nil}
-
-  def cast({:maybe, type}, value) do
+  @spec cast(t, term) :: {:ok, term} | {:ok, term, boolean} | {:error, keyword()} | :error
+  def cast(type, value, current \\ nil)
```

Comment:
As I mentioned in the previous discussion, we need to introduce a new `cast` function. We cannot repurpose this one because we do not always have a `current` and assuming that `current` is nil will lead to the wrong behaviour in places like using `cast` to cast a query.

Here is a more concrete proposal: what if introduce a new function and call it `cast_change`? It can be a contrast to `put_change`, which @Eiji7 is working on.

Reply:
@josevalim Whatever you decide. However here are my 2 cents:

`cast_change` definitely would be confusing as `cast` was previously for `validation` and `change` (per documentation) for internal application data only.

> Note the value is directly stored in the changeset with no validation whatsoever. For this reason, this function is meant for working with data internal to the application.

Also I wonder if we even need `change/2` callback and simply add support for `parameterized` inside `Ecto.Changeset.change/2`, so the only callback parameterized types needs to implement would be `cast`.

Thoughts?

Reply:
The issue is that cast and change for embeds behave very differently. Furthermore, cast with current and without current also behave differently. So we do need all of them as distinct things, the question is which ones we want to expose.

Reply:
Hmm … `cast` for me is obvious, but as `change` documentation says we should just put new data without any validation (so I suppose also no `raise` for `:on_replace` option etc.) as this is for internal application data only, so developer should know what he/she is changing. If I would like to let's say update (in context of `embeds`) I would simply use `Map.merge/2` before even calling `change`. Of course that's a breaking change, but if we are going with "new" `Ecto.Type` (i.e. `Ecto.ParameterizedType`) we can always add something like `Ecto.ParameterizedEmbed` just to make things simpler if current implementation is too complicated …

Reply:
If we don't want to support embeds (or we want to special case anyway), then I would just rather not do any of this, since the whole point of adding these new callbacks is to unify embeds with parameterized types.

Reply:
@josevalim ok, now after I look at new code I see how `cast_change` is going work … I can see that 

I only wonder why `cast` is always raising on `cast/3` :thinking: 

```elixir
  @impl Ecto.ParameterizedType
  def cast(_, %{field: field, owner: owner}) do
    raise "cannot cast embed `#{field}` from `#{owner}` in queries"
  end
```

Should I wait for this PR and do same for `change` or maybe I'm just confused with `cast_change` name and `change` itself should do something else?

Reply:
> change itself should do something else?

`change` does something else, so no worries!